refactor/ast.py

```diff
@@ -32,17 +35,28 @@ def join(self) -> str:
         """Return the combined source code."""
         return "".join(map(str, self.lines))
 
-    def apply_indentation(
+    def apply_source_formatting(
         self,
-        indentation: StringType,
+        source_lines: Lines,
         *,
-        start_prefix: AnyStringType = "",
-        end_suffix: AnyStringType = "",
+        markers: Tuple[int, int, int | None] = None,
     ) -> None:
-        """Apply the given indentation, optionally with start and end prefixes
-        to the bound source lines."""
+        """Apply the indentation from source_lines when the first several characters match
+
+        :param source_lines: Original lines in source code
+        :param markers: Indentation and prefix parameters. Tuple of (start line, col_offset, end_suffix | None)
+        """
+
+        def not_original(i: int) -> bool:
+            common_chars: str = commonprefix([str(self.data[i]), str(source_lines.data[i].data)])
+            is_multiline_string: int = str(self.data[i]).find(common_chars) == 0 and common_chars in ["'''", '"""']
+            return not (i < len(source_lines.data) and (str(self.data[i]) == common_chars or is_multiline_string))
```

Comment:
So if we reduce it to bare minimum (let's drop the multiline strings for now, just to make this patch simpler), what this does is basically checking if the original line starts with the replaced line and if it is the case we skip indenting?

A better way to express it might be this:
```py
        for index, line in enumerate(self.data):
            if index < len(source_lines):
                original_line = source_lines[index]
            else:
                original_line = None

            if index == 0:
                self.data[index] = indentation + str(start_prefix) + str(line)  # type: ignore
            elif original_line is not None and original_line.startswith(line):
                self.data[index] = line # type: ignore
            else:
                self.data[index] = indentation + line  # type: ignore
```

But I am still unsure on how this should work in different scenarios, would love to see a few more examples (or actually tests) where the indentation is preserved for the first argument (when replacing the whole call), the middle arguments, the last argument (maybe introduce a new argument or remove one from the middle).

Reply:
It only works when the `Call` is unchanged, removing or adding arguments does not maintain the original form, but collapses the `Call` into one line. The case I was working with was wrapping the `Call` in `Await` and it works well in that case

There is something I don't understand though, I added a few tests (bear with my padawan-level) and the Issue #12 pass, but in the complete_rules it does not, it'd be great if you took a look at those tests, I must be doing something (or more) incorrectly

Reply:
Ok, the issue was that in the `test_ast.py`, I use the full source_lines (with indent=0 and thus is not added to the failing line), but in the `test_complete_rules.py` only the changed source line is passed (because of the `view`), so the indentation is there and the test fails.
Now, the 'multi-line' issue is more of a "within parens" issue, so would it be acceptable to just have this:
`elif original_line is not None and (original_line.startswith(line) or line[-1] in (")}]")):`
This way the test passes